auth/src/main/java/org/aerogear/auth/impl/OIDCAuthCodeImpl.java

```diff
@@ -20,6 +43,120 @@ public OIDCAuthCodeImpl(final AuthServiceConfig config) {
+        if (userIdentity != null) {
+            Map<RoleKey, IRole> realmRoles = parseRealmRoles();
+            if (realmRoles != null) {
+                roles.putAll(realmRoles);
+            }
+            Map<RoleKey, IRole> clientRoles = parseClientRoles();
+            if (clientRoles != null) {
+                roles.putAll(clientRoles);
+            }
+        }
+        return roles;
+    }
+
+    private Map<RoleKey, IRole> parseRealmRoles() throws JSONException {
+        Map<RoleKey, IRole> realmRoles = new HashMap<>();
+        if (userIdentity.has(REALM) && userIdentity.getJSONObject(REALM).has(ROLES)) {
+            String tokenRealmRolesJSON = userIdentity.getJSONObject(REALM).getString(ROLES);
+
+            String realmRolesString = tokenRealmRolesJSON.substring(1, tokenRealmRolesJSON.length() - 1);
+            String roles[] = realmRolesString.split(COMMA);
+
+            for (String rolename : roles) {
+                RealmRole realmRole = new RealmRole(rolename);
+                realmRoles.put(new RoleKey(realmRole), realmRole);
+            }
+        }
+        return realmRoles;
+    }
+
+    private Map<RoleKey, IRole> parseClientRoles() throws JSONException {
+        Map<RoleKey, IRole> clientRoles = new HashMap<>();
+
+        AuthServiceConfig authConfig = this.getConfig();
+        JSONObject authJSON =  authConfig.toJSON();
+
+        if (authJSON.has(RESOURCE)) {
+            String initialClientID = authJSON.getJSONObject(RESOURCE).toString();
+
+            if (userIdentity.has(CLIENT) && userIdentity.getJSONObject(CLIENT).has(initialClientID)
+                    && userIdentity.getJSONObject(CLIENT).getJSONObject(initialClientID).has(ROLES)) {
+                String tokenClientRolesJSON = userIdentity.getJSONObject(CLIENT).getJSONObject(initialClientID).getString(ROLES);
+
+                String clientRolesString = tokenClientRolesJSON.substring(1, tokenClientRolesJSON.length() - 1);
+                String roles[] = clientRolesString.split(COMMA);
+
+                for (String rolename : roles) {
+                    ClientRole clientRole = new ClientRole(rolename);
+                    clientRoles.put(new RoleKey(clientRole), clientRole);
+                }
+            }
+        }
+        return clientRoles;
+    }
+
+    private JSONObject getIdentityInformation(final ICredential credential) throws JSONException {
+        String accessToken = credential.getAccessToken();
+        JSONObject decodedIdentityData = new JSONObject();
+
+        try {
+            // Decode the Access Token to Extract the Identity Information
+            String[] splitToken = accessToken.split("\\.");
+            byte[] decodedBytes = Base64.decode(splitToken[1], Base64.URL_SAFE);
+            String decoded = new String(decodedBytes, "UTF-8");
+            try {
+                decodedIdentityData = new JSONObject(decoded);
+            } catch (JSONException e) {
+                e.printStackTrace();
```

Comment:
Throw an exception here, passing the JSONException as root exception.
There is an AuthenticationException that could be used here.